src/diff_edit.rs

```diff
@@ -96,6 +137,111 @@ fn set_readonly_recursively(path: &Path) -> Result<(), std::io::Error> {
     }
 }
 
+pub fn run_mergetool(
+    _ui: &mut Ui,
+    tree: &Tree,
+    repo_path: &RepoPath,
+) -> Result<TreeId, ConflictResolveError> {
+    let conflict_id = match tree.path_value(repo_path) {
+        Some(TreeValue::Conflict(id)) => id,
+        Some(_) => return Err(ConflictResolveError::NotAConflictError(repo_path.clone())),
+        None => return Err(ConflictResolveError::PathNotFoundError(repo_path.clone())),
+    };
+    let conflict = tree.store().read_conflict(repo_path, &conflict_id)?;
+    let mut content = match extract_file_conflict_as_single_hunk(tree.store(), repo_path, &conflict)
+    {
+        Some(c) => c,
+        _ => {
+            let mut summary_bytes: Vec<u8> = vec![];
+            describe_conflict(&conflict, &mut summary_bytes)
+                .expect("Writing to an in-memory buffer should never fail");
+            return Err(ConflictResolveError::NotNormalFilesError(
+                repo_path.clone(),
+                String::from_utf8_lossy(summary_bytes.as_slice()).to_string(),
+            ));
+        }
+    };
+    // The usual case is 1 `removes` and 2 `adds`. 0 `removes` means the file did
+    // not exist in the conflict base. Only 1 `adds` may exist for an
+    // edit-delete conflict.
+    if content.removes.len() > 1 || content.adds.len() > 2 {
+        return Err(ConflictResolveError::ConflictTooComplicatedError {
+            path: repo_path.clone(),
+            removes: content.removes.len(),
+            adds: content.adds.len(),
+        });
+    };
+
+    let mut materialized_conflict: Vec<u8> = vec![];
+    materialize_merge_result(&content, &mut materialized_conflict)
+        .expect("Writing to an in-memory buffer should never fail");
+    let materialized_conflict = materialized_conflict;
+
+    let temp_dir = tempfile::Builder::new()
+        .prefix("jj-resolve-")
+        .tempdir()
+        .map_err(ExternalToolError::SetUpDirError)?;
+    let files: HashMap<&str, _> = [
+        ("base", content.removes.pop().unwrap_or_default()),
+        ("right", content.adds.pop().unwrap_or_default()),
+        ("left", content.adds.pop().unwrap_or_default()),
+        ("output", materialized_conflict.clone()),
+    ]
+    .into_iter()
+    .collect();
+    let suffix = match repo_path.components().last() {
+        Some(filename) => format!("_{}", filename.as_str()),
+        None => String::default(),
+    };
```

Comment:
or `.unwrap()`? I don't think root `repo_path` is allowed here.

Then, this `format!()` can be inlined into `temp_dir.path().join(format!("{role}{suffix}"))`.

Reply:
I can't imagine a situation when the root repo path would appear here, but if this changes in the future, I don't want a crash because of such a minor issue.